data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,544 @@
+
+For one, the types are lost at execution time. That means that the type's structure
+cannot be deconstructed at execution time to control the flow. This is
+why no general `to_string : 'a -> string` or `print : 'a -> ()` function can
+exist in compiled binaries (in the toplevel, types are kept).
+
+So any general function that depends on the type's structure _must_ be
+written at compile time, when types are still available.
+
+Secondly, one of the strong features of OCaml is its type system, which can be
+used to check the sanity of many things. An example of this is the [`Tyxml`
+library](https://github.com/ocsigen/tyxml), which uses the type system to ensure
+that the resulting HTML value satisfies most of the W3C standards. However,
+writing `Tyxml` values can be cumbersome compared to writing HTML syntax.
+Transforming HTML code into OCaml values at compile time allows users to keep both the
+type-checker on the produced value and the familiarity of writing HTML code.
+
+Other rewriters, such as `ppx_expect`, show that being able to enrich the syntax
+via PPX rewriters is very useful, even outside of the specificity of OCaml.
+
+## The Need for Controlling the PPX Ecosystem: `ppxlib`
+
+Although PPXs are great for generating code at compile time, they raise a few
+questions, especially in the presence of multiple PPX rewriters.
+- What is the semantics of the composition of multiple PPXs? The order might
+  matter.
+- How can I trust that some part of my code is not modified when using
+  third-party PPXs?
+- How can I keep compilation time short when using many rewriters?
+- How can I write a PPX easily if I have to deal with parsing or demarshalling
+  an AST?
+- How can I deal with such a [long and complex
+  type](https://v2.ocaml.org/api/compilerlibref/Parsetree.html) as in Parsetree?
+- How can I solve the problem that new OCaml versions tend to add new features
+  to the language and therefore need to enrich and break the Parsetree types?
+
+Many of these questions stem from the fact that there is no macro language part
+of OCaml and that the preprocessors are always standalone programs. This means
+that they can do anything (while macros usually restrict the expressivity of the
+preprocessing), and the compiler has no control over them. However, the OCaml
+Platform includes a library to write PPXes, that somehow act as an macro
+language without losing the full generality of PPXs:
+[ppxlib](https://github.com/ocaml-ppx/ppxlib/). This library provides generic
+ways of writing [extenders](#extension-nodes-and-extenders) and
+[derivers](#attributes-and-derivers), ensuring that they will work well
+together, removing the composition problem we had with multiple arbitrary
+transformations. It also outputs one single binary, even in the case of multiple
+transformation registered.
+
+With `ppxlib`, PPX authors can concentrate on their own part—the rewriting
+logic. Then, they can register their transformation, and `ppxlib` will be
```

Comment:
```suggestion
With `ppxlib`, PPX authors can individually concentrate on the rewriting
logic. Then, they can register their transformation, and `ppxlib` will be
```
Does this convey the intention?
...or... "PPX authors can concentrate on their own part: the rewriting logic."